lib/mix/lib/mix/config.ex

```diff
@@ -155,10 +163,18 @@ defmodule Mix.Config do
   `file` is the path to the configuration file to be read. If that file doesn't
   exist or if there's an error loading it, a `Mix.Config.LoadError` exception
   will be raised.
+
+  `loaded_paths` is a list of configuration files that have been previously
+  read. If `file` exists in `loaded_paths`, a `Mix.Config.LoadError` exception
+  will be raised.
   """
-  def read!(file) do
+  def read!(file, loaded_paths \\ []) do
     try do
-      {config, binding} = Code.eval_file(file)
+      if file in loaded_paths do
+        raise ArgumentError, message: "Recursive load of #{file} detected."
```

Comment:
This looks great! My only suggestion is to not change the API of
Code.eval_file. Instead, call File.read! and Code.eval_string.

On Monday, June 13, 2016, Andrea Leopardi <notifications@github.com> wrote:

> In lib/mix/lib/mix/config.ex
> <https://github.com/elixir-lang/elixir/pull/4809#discussion_r66728456>:
>
> >      try do
> > -      {config, binding} = Code.eval_file(file)
> > +      if file in loaded_paths do
> > +        raise ArgumentError, message: "Recursive load of #{file} detected."
>
> We use lowercase first letter and no final period for raised messages
> (except Mix.raise) :)
>
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/elixir-lang/elixir/pull/4809/files/32d0301eaadf5e969781e4fab1dacbfdcd03b42f#r66728456>,
> or mute the thread
> <https://github.com/notifications/unsubscribe/AAAlbpdUU9npDbe-khlO1q7m96xwR2IFks5qLIJqgaJpZM4Iz5NQ>
> .
>


-- 


*José Valim*
www.plataformatec.com.br
Skype: jv.ptec
Founder and Director of R&D